admin/sales_analytics/metrics.py

```diff
@@ -0,0 +1,113 @@
+    """
+    start_date = datetime.now() - timedelta
+    pipeline = [
+        {'$match': {'date': {'$gt': start_date}}},
+        {'$group': {'_id': '$user', 'node_id': {'$addToSet': '$params.node'},
+                    'action': {'$addToSet': '$action'}}}
+    ]
+    user_node = db.nodelog.aggregate(pipeline)['result']
+    multi_product_count = 0
+    cross_product_count = 0
+    multi_action_count = 0
+    for i in user_node:
+        if i['_id']:
+            user_id = i['_id']
+            node_id = i['node_id']
+            product = []
+            nodes = db.node.find({'_id': {'$in': node_id}})
+            for node in nodes:
+                product.append(get_entry_point(node['system_tags']))
+            if len(set(product)) > 1:
+                multi_product_count += 1
+
+            # Cross product count
+            user = db.user.find_one({'_id': user_id})
+            user_entry_point = get_entry_point(user['system_tags'])
+            for j in product:
+                if user_entry_point == product:
+                    cross_product_count += 1
+
+            # Action type
+            if len(set(i['action'])) > 1:
+                multi_action_count += 1
+
+    return {'multi_product_count': multi_product_count,
+            'cross_product_count': cross_product_count,
+            'multi_action_count': multi_action_count,
+            }
+
+
+def get_repeat_action_user_count(db=db, timedelta=timedelta(days=30)):
+    """
+    Get the number of users that have repetitive actions (with a 3 second difference)
+    during the last month.
+    """
+    start_date = datetime.now() - timedelta
+    pipeline = [
+        {'$match': {'date': {'$gt': start_date}}},
+        {'$group': {'_id': '$user', 'nodelog_id': {'$addToSet': '$_id'}}},
+    ]
+
+    user_nodelog = db.nodelog.aggregate(pipeline)['result']
+    repeat_action_count = 0
+    repeat_action_user_age = []
+    for i in user_nodelog:
+        if i['_id']:
+            user_id = i['_id']
+            nodelog_id = i['nodelog_id']
+            nodelogs = db.nodelog.find({'_id': {'$in': nodelog_id}}).sort([('date', 1)])
+            repeat_action_date = {}
+            for nodelog in nodelogs:
+                action = nodelog['action']
+                date = nodelog['date']
+                if action not in repeat_action_date:
+                    repeat_action_date[action] = date
+                elif abs((date - repeat_action_date[action]).total_seconds()) < 3:
+                    repeat_action_date[action] = date
+                else:
+                    repeat_action_count += 1
+                    date_registered = db.user.find({'_id': user_id}).next()['date_registered']
+                    age = (date - date_registered).days
+                    repeat_action_user_age.append(age)
+                    break
+    return {'repeat_action_count': repeat_action_count, 'repeat_action_age': repeat_action_user_age}
+
+
+user_count = get_user_count(db, ENTRY_POINTS)
```

Comment:
Those variables are used by views.py to be put in context. This way I could import variables in views.py instead of the functions and all the variables needed to pass to the functions. I thought this makes views.py much cleaner.